crates/libcontainer/src/container/container_delete.rs

```diff
@@ -58,20 +92,22 @@ impl Container {
                             .with_context(|| "failed to run post stop hooks")?;
                     }
                 }
-
-                // remove the directory storing container state
-                log::debug!("remove dir {:?}", self.root);
-                fs::remove_dir_all(&self.root).with_context(|| {
-                    format!("failed to remove container dir {}", self.root.display())
-                })?;
+                Err(err) => {
+                    // There is a brief window where the container state is
+                    // created, but the container config is not yet generated
+                    // from the OCI spec. In this case, we assume as if we
+                    // successfully deleted the config and moving on.
+                    log::warn!("skipping loading youki config due to: {err:?}, continue to delete");
+                }
```

Comment:
Do we need to worry about race condition? For eg, if we are deleting a container that is being created, then there might be a window where the config is not generated, and we are treating it as deleted. If that does get created after this moment, then should we be concerned about it or not?

If we are treating the delete root after this as catch all, then I guess this is fine

Reply:
You asked a very good question and here is what I think. First of all, the actual bug is addressed in (https://github.com/containers/youki/pull/1649) and this PR only adds a warning and an explanation, so the context of the bug is in #1649 and #1640. In short, if rootless::New errors for whatever reason, we are stuck in a state where container state is created but not the youki config. This will cause the delete to not able to clean up because it is missing the youki config.

Here is the complex part about the race condition and you are right that there is a potential race here. If process 1 is calling `youki create` and process 2 is calling `youki delete --force`, we have a window where process 1 has not launched the process yet, but process 2 is going ahead to delete. There is no way for process 2 to signal process 1 to stop creating the container processes with force flag. What will happen is that process 2 will delete the container root and process 1 will continue to start the process wand wait on the exec.fifo, which is a file (named pipe). So when container root is deleted, the exec.fifo will be gone and the process waiting on it will error out. Therefore, both should be clean up correctly. Or process 1 at any point trying to save the state and fails. So in theory, the race condition will not cause leakage in terms of process or files on the system. I believe this the best we can guarantee. I think this covers all the situations, but please double check my work here.

Reply:
Hey, I think this makes sense :+1:
I think this should solve the issue and not cause any leakage of resources.
If possible, give me a day's time, I'll try to do a manual verification check by launching parallel processes and checking that resources are cleaned up.

In either case, I think this implementation is good enough :)

Reply:
> Hey, I think this makes sense 👍 I think this should solve the issue and not cause any leakage of resources. If possible, give me a day's time, I'll try to do a manual verification check by launching parallel processes and checking that resources are cleaned up.
> 
> In either case, I think this implementation is good enough :)

Take you time please :) I am curious as well.

Reply:
Hey @yihuaf Unfortunately, I didn't get time in last week to do this :sweat:  I'll try to get this done soon :pray:

Reply:
Another option is to merge this PR as is, since the PR is fixing a specific issue. We can continue to test the edge case on main and file issue as needed.